src/commands/mod.rs

```diff
@@ -736,6 +774,372 @@ impl DaemonControl {
+        let mempool_entry = self
+            .bitcoin
+            .mempool_entry(txid)
+            .ok_or(CommandError::RbfError(RbfErrorInfo::TxNotYetBroadcast(
+                *txid,
+            )))?;
+        // Check sum of descendant counts of all directly conflicting transactions is not too high.
+        // We are assuming we only have one directly conflicting transaction.
+        if mempool_entry.descendant_count > RBF_MAX_DESCENDANT_COUNT {
+            return Err(CommandError::RbfError(RbfErrorInfo::TooManyDescendants(
+                mempool_entry.descendant_count,
+            )));
+        }
+        let prev_feerate_vb = {
+            let prev_abs_fee = prev_psbt
+                .fee()
+                .expect("fee must be Ok as transaction is already in mempool");
+            let prev_tx_size: u64 = prev_psbt
+                .clone()
+                .extract_tx()
+                .vsize()
+                .try_into()
+                .expect("vsize must fit in a u64");
+            prev_abs_fee
+                .checked_div(prev_tx_size)
+                .ok_or(CommandError::InsaneFees(InsaneFeeInfo::InvalidFeerate))?
+        };
+        // Check replacement transaction's feerate is greater than previous feerate (the only directly conflicting transaction).
+        // TODO: Should we throw error or just set feerate_vb = prev_feerate_vb + 1 and check `rbf_psbt` below?
+        if feerate_vb < prev_feerate_vb.to_sat().checked_add(1).unwrap() {
+            return Err(CommandError::RbfError(RbfErrorInfo::TooLowFeerate(
+                feerate_vb,
+            )));
+        }
+        let prev_txos = &prev_psbt.unsigned_tx.output;
+        let mut destinations: HashMap<bitcoin::Address, bitcoin::Amount> =
+            HashMap::with_capacity(prev_txos.len());
+        let mut change_index: Option<bip32::ChildNumber> = None;
+        for txo in prev_txos {
+            let address = bitcoin::Address::from_script(
+                &txo.script_pubkey,
+                self.config.bitcoin_config.network,
+            )
+            .map_err(CommandError::Address)?;
+            match db_conn.derivation_index_by_address(&address) {
+                Some((ind, true)) => {
+                    // TODO: check if there is more than one change output?
+                    println!("found change at address {} and index {}", address, ind);
+                    change_index = Some(ind);
+                }
+                _ => {
+                    // Non-change outputs are handled in the same way whether they are ours or not.
+                    if !is_cancel {
+                        destinations.insert(address, bitcoin::Amount::from_sat(txo.value));
+                    }
+                }
+            };
+        }
+        // We take all previous inputs as mandatory candidates and, if not self-send, include
+        // confirmed coins as optional.
+        // TODO: For `is_cancel`, shall we include just one previous input?
```

Comment:
OK yep that makes sense. I think starting with the largest would be good.

Reply:
I've changed the `is_cancel` case so that in the first instance, it uses the previous coins as optional candidates and we let the coin selection decide which of them are best to use given the min fee and feerate. If there are insufficient funds using just the previous coins as candidates, it then uses the same approach as for `!is_cancel`, i.e. all previous coins are mandatory candidates and confirmed coins are optional.

Reply:
This approach for `is_cancel` using optional candidates doesn't always work. The replacement transaction is a self-send, but it can happen if we have optional candidates that coin selection gives us a changeless solution without the optional candidates rather than one with change using the optional candidates. I think I'll need to use just mandatory candidates, adding one at a time as originally suggested until we have a solution with change.

Reply:
I think this could also affect the case where we want to bump the fee of a self-send: the higher fee may lead to a solution with no change. I'll use a similar approach for that case too.

One of the available scoring metrics for coin selection is [`Changeless`](https://github.com/evanlinjin/bdk/blob/d620dc61e809a8c280d623d5275abe8413b4b00a/nursery/coin_select/src/metrics/changeless.rs#L5). In principle, we might be able to use a similar `WithChange` metric by editing this [line](https://github.com/evanlinjin/bdk/blob/d620dc61e809a8c280d623d5275abe8413b4b00a/nursery/coin_select/src/metrics/changeless.rs#L18) to end with `is_some()` instead of `is_none()`. Then we could just pass the mandatory and optional candidates without having to increment the number of candidates.

Reply:
I did a quick test using a modified version of the `LowestFee` metric:
```
pub struct LowestFeeWithChange<'c, C> {
    pub lowest_fee: LowestFee<'c, C>, // This is the metric we're currently using
    pub must_have_change: bool,
}
```
 so that its score is given by:
```
if cs.is_target_met(self.lowest_fee.target, drain)
    && (drain.is_some() | !self.must_have_change)
{
    self.lowest_fee.score(cs) // We reuse the `LowestFee` score function.
} else {
    None
}
```
So `LowestFeeWithChange` is basically the same as `LowestFee`, except that `score()` will return `None` in case there's no change (depending on `must_have_change`).

From my tests, this avoids the problems I was facing with changeless solutions and also avoids the need for the extra logic adding candidates one at a time.

Reply:
Hmm I think it's a bit more complicated as we would also need to change how `bound()` is implemented. I'll continue with the iterating approach for now, adding candidates one by one.

Reply:
Sorry for all the noise, but I think this approach is still working. I just had to apply the same logic to the fallback in case BnB gave no solution. I've made the changes in a separate branch:
https://github.com/wizardsardine/liana/compare/master...jp1ac4:liana:lowest-fee-with-change-metric

It may be worth considering this approach as it should be simpler than iterating over the set of candidates as well as the min fee.

Reply:
This (the code and the explanation) makes sense to me.

It reminds me of the case where an output has more than one output paying to one of our addresses. Another approach could be to just set the output(s) as part of the destination, but it's ugly as it might create another change output while there is already one (or more). On the other hand it would conserve the value of the output(s). Can this ever be a reasonable expectation of the user? Hmm i don't think it'd be worth risking further duplication of the change.
TL;DR: i think you should go with your with-change metric.

Reply:
I also considered an approach of treating the change as a regular destination in order to avoid a changeless solution, but had similar concerns.

OK great, I'll add the with-change metric in a separate PR.

Reply:
PR for new metric: https://github.com/wizardsardine/liana/pull/832.